tests/e2escenarios/e2e_test.go

```diff
@@ -270,4 +298,110 @@ var _ = Describe("E2E Test", func() {
+			Expect(addBindableKind.ExitCode()).To(BeEquivalentTo(0))
+		})
+
+		It("should verify developer workflow of using binding as env in innerloop", func() {
+			bindingName := helper.RandString(6)
+
+			command := []string{"odo", "init"}
+			_, err := helper.RunInteractive(command, nil, func(ctx helper.InteractiveContext) {
+
+				// helper.ExpectString(ctx, "Based on the files in the current directory odo detected")
+				helper.ExpectString(ctx, "Language: Go")
+				helper.ExpectString(ctx, "Project type: Go")
+				helper.ExpectString(ctx, "Is this correct")
+
+				helper.SendLine(ctx, "")
+
+				helper.ExpectString(ctx, "Select container for which you want to change configuration?")
+
+				helper.SendLine(ctx, "")
+
+				helper.ExpectString(ctx, "Enter component name")
+
+				helper.SendLine(ctx, componentName)
+
+				helper.ExpectString(ctx, "Your new component '"+componentName+"' is ready in the current directory")
+
+			})
+			Expect(err).To(BeNil())
+			Expect(helper.ListFilesInDir(commonVar.Context)).To(ContainElement("devfile.yaml"))
+
+			// // "execute odo dev and add changes to application"
+			var devSession helper.DevSession
+			var ports map[string]string
+
+			devSession, _, _, ports, err = helper.StartDevMode(nil)
+			Expect(err).ToNot(HaveOccurred())
+
+			// "send data"
+			_, err = receiveData(fmt.Sprintf(ports["8080"] + "/api/user"))
+			Expect(err).ToNot(BeNil()) // should fail as application is not connected to DB
+
+			//add binding information (binding as ENV)
+			helper.Cmd("odo", "add", "binding", "--name", bindingName, "--service", "cluster-example-initdb", "--bind-as-files=false").ShouldPass()
+
+			// Get new random port after restart
+			Eventually(func() map[string]string {
+				_, _, ports, err = devSession.GetInfo()
+				Expect(err).ToNot(HaveOccurred())
+				return ports
+			}, 180, 10).ShouldNot(BeEmpty())
+
+			// "send data"
+			data := sendDataEntry(ports["8080"])
+			Expect(data["message"]).To(Equal("User created successfully"))
+
+			// "get all data"
+			rec, err := receiveData(fmt.Sprintf(ports["8080"] + "/api/user"))
+			Expect(err).To(BeNil())
+			helper.MatchAllInOutput(rec, []string{"id", "1", "name", "joe", "location", "tokyo", "age", "23"})
+
+			// check odo describe to check for env
+			stdout := helper.Cmd("odo", "describe", "binding").ShouldPass().Out()
+			helper.MatchAllInOutput(stdout, []string{"Available binding information:", "CLUSTER_HOST", "CLUSTER_PASSWORD", "CLUSTER_USERNAME"})
+
+			// "running odo list"
+			stdout = helper.Cmd("odo", "list").ShouldPass().Out()
+			helper.MatchAllInOutput(stdout, []string{componentName, "Go", "Dev", bindingName})
+
+			// "exit dev mode"
+			devSession.Stop()
+			devSession.WaitEnd()
+
+			// remove bindings and check devfile to not contain binding info
+			// TODO: move `remove binding` inside devsession after https://github.com/redhat-developer/odo/issues/6101 is fixed
+			helper.Cmd("odo", "remove", "binding", "--name", bindingName).ShouldPass()
```

Comment:
#6101 has been fixed.

Reply:
As this PR has been around for so many months, I think it is okay to merge it with the `TODO` note. If we want, we can create a separate PR to change the behavior here with #6101 in mind. WDYT?